### PR TITLE
api shouldn't error when xform is stored in sql

### PR DESCRIPTION
**Pull request by nickpell**
http://manage.dimagi.com/default.asp?232577

### CODE REVIEW

**Comment by nickpell** [comment]
@kaapstorm

File: corehq/apps/api/resources/v0_4.py [hunk 1]
```diff
@@ -120,6 +120,7 @@ def obj_get_list(self, bundle, domain, **kwargs):
 
         def wrapper(doc):
             if doc['doc_type'] in xform_doc_types:
+                doc.pop('user_id', None)  # needed when xform is stored in sql
```
Comment (czue) [inline corehq/apps/api/resources/v0_4.py]: this seems like an ok workaround but i'm curious why we need to wrap these objects at all?

Comment (nickpell) [inline corehq/apps/api/resources/v0_4.py]: not sure, I didn't dig too deeply.

File: corehq/apps/api/resources/v0_4.py [hunk 2]
```diff
@@ -120,6 +120,7 @@ def obj_get_list(self, bundle, domain, **kwargs):
...
                 return xform_doc_types[doc['doc_type']].wrap(doc)
```
Comment (benrudolph) [inline corehq/apps/api/resources/v0_4.py]: i think this is more thoroughly broken then we think. right now what's happening is when the sql form gets moved to json, it adds a doc_type property that tricks this "wrapper" into thinking it's couch and then we try and wrap the sql doc in a couch class. XFormInstanceSQL doesn't have a wrap class

Comment (nickpell) [inline corehq/apps/api/resources/v0_4.py]: I agree with that reasoning.

Comment (czue) [inline corehq/apps/api/resources/v0_4.py]: @benrudolph I think the `XFormInstanceSQLSerializer` intentionally sets `'doc_type'` to `'XFormInstance'` so that couch and sql are consistent so I don't think that's happening. I still don't get why we would need to wrap the json in the APIs though. Guessing just wanting to take advantage of some computed properties which we could maybe get rid of by making the serialization from sql/couch more explicit.

Comment (czue) [inline corehq/apps/api/resources/v0_4.py]: see https://github.com/dimagi/commcare-hq/blob/master/corehq/form_processor/models.py#L266 and https://github.com/dimagi/commcare-hq/blob/master/corehq/form_processor/serializers.py#L36

Comment (benrudolph) [inline corehq/apps/api/resources/v0_4.py]: agree with the consistency bit, but this bit looks suspricious:

```
xform_doc_types[doc['doc_type']].wrap(doc)
```
isn't that saying wrap a sql form as a couch form?

agree probably best not to wrap

Comment (czue) [inline corehq/apps/api/resources/v0_4.py]: wow, sorry i totally misread your comment. yeah we're saying the same thing.

Comment (kaapstorm) [inline corehq/apps/api/resources/v0_4.py]: If I'm understanding the history of this correctly, I can imagine why this might be useful; You want to duck-type SQL instances as Couch instances, e.g. to use [over here][1]. Deserializing from JSON using an API SQL serializer might be a better approach, but it can be annoying because it tries to save the instance (which already exists). To avoid the save, you can do something like this (effectively a hacky "wrap" for SQL classes): 

```python
def get_instance_from_data(SerializerClass, data):
    """
    Return a deserialized instance from serialized data

    cf. https://github.com/tomchristie/django-rest-framework/blob/master/rest_framework/serializers.py#L71
    and https://github.com/tomchristie/django-rest-framework/blob/master/rest_framework/serializers.py#L882
    """
    # It does seem like you should be able to do this in one line. Django REST 
    # framework assumes that you always want to `save()`. This function avoids 
    # saving.
    ModelClass = SerializerClass.Meta.model
    serializer = SerializerClass(data=data)
    assert serializer.is_valid(), 'Unable to deserialize data while creating {}'.format(ModelClass)
    # ^^^ We need to call `is_valid()` to get to `validated_data` vvv
    return ModelClass(**serializer.validated_data)
```

The alternative is to wrap it as a Couch instance -- i.e. what I think @snopoke seems to have done here -- and which might turn out to be the better of the alternatives.


  [1]: https://github.com/dimagi/commcare-hq/blob/03c98ab17d6b8180ab638ffb7e0e520aec3b25e7/corehq/apps/api/es.py#L564-L564

Comment (snopoke) [inline corehq/apps/api/resources/v0_4.py]: This wasn't added by me. Originally added here: https://github.com/dimagi/commcare-hq/commit/d3f1c857d5721e5081b44f1d63dde17db1f3baa7 when we started using ES as the API backend (perviously we didn't have a 'form list' API).

The reason for the wrapping is that Tastypie is designed to work off model instances. After looking at this a bit more I suspect the API's are broken in even more ways e.g. I doubt this works: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/api/resources/v0_4.py#L64-L69

Comment (benrudolph) [inline corehq/apps/api/resources/v0_4.py]: fwiw, i'd prefer to avoid wrapping a sql xform instance as a couch instance as it could cause some potentially very confusing errors in the future. the couch instance adds some methods that should never be called on a sql form, such `.save`. if that were called then we'd end up with a sql xform and a duplicate couch xform. other methods would also be broken, but probably have less ramifications.

it's come up before that we needed to be able to wrap a sql xform, this might be the opportunity to try and make one of those kinds of functions. don't know enough about tastypie to know if there's a workaround

Comment (snopoke) [inline corehq/apps/api/resources/v0_4.py]: One option is to do a non-backwards compatible v0_6 using Django Rest
Framework (just for forms and cases). Also looks like tasypie has support
for non ORM data:
http://django-tastypie.readthedocs.io/en/latest/non_orm_data_sources.html.

Also seems like if we can avoid wrapping them at all that would be a big
win. And I don't really see why we shouldn't be able to. We've already put
effort into standardizing the ES data format so we should just be able to
serve a transformed version of that (cases may require more transformation
than forms if we want to stick with the current API format).

Comment (benrudolph) [inline corehq/apps/api/resources/v0_4.py]: +1 for seeing how hard it would be to write a rest framework api

Comment (esoergel) [inline corehq/apps/api/resources/v0_4.py]: Who owns this issue at the moment?  Is it the case that the forms api doesn't work at all for sql domains?  Is anyone working on a new API for SQL backed forms?

Comment (benrudolph) [inline corehq/apps/api/resources/v0_4.py]: haha that's a good question, i had originally thought that this was being handled by interrupt, but it seems that that's not the case. it may be worth putting back on the scale agenda and having it resourced through scale

Comment (benrudolph) [inline corehq/apps/api/resources/v0_4.py]: we had done a pass to convert the apis to be sql ready, but seems like there's some more work to do

Comment (benrudolph) [inline corehq/apps/api/resources/v0_4.py]: (putting it on scale agenda and therefore declaring that scale owns it since it seems out of scope for interrupt)
